### PR TITLE
ENH: Styler tooltips feature

### LINKED ISSUE #21266: [Enhancement Request] Styler support for tooltips.
#### Problem description
It would be nice if pandas stlying allowed you to set a tooltip for table cells, in the same way that it allows you to set the CSS style using apply, applymap, etc.

Currently it is possible to create a hacky workaround using css-only tooltips and set_table_styles, but it adds substantial complexity to achieve something that is conceptually very simple.

#### Output of ``pd.show_versions()``

<details>

[paste the output of ``pd.show_versions()`` here below this line]
pd.show_versions()

INSTALLED VERSIONS
------------------
commit: None
python: 2.7.13.final.0
python-bits: 64
OS: Linux
OS-release: 4.9.0-6-amd64
machine: x86_64
processor: 
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8
LOCALE: None.None

pandas: 0.19.2
nose: None
pip: 9.0.1
setuptools: 36.0.1
Cython: 0.28.3
numpy: 1.14.3
scipy: 0.19.1
statsmodels: 0.8.0
xarray: None
IPython: 5.1.0
sphinx: None
patsy: 0.4.1+dev
dateutil: 2.6.1
pytz: 2017.2
blosc: None
bottleneck: None
tables: 3.4.2
numexpr: 2.6.2
matplotlib: 2.0.0
openpyxl: 2.3.0
xlrd: 1.1.0
xlwt: 0.7.5
xlsxwriter: None
lxml: 4.0.0
bs4: 4.6.0
html5lib: 0.999999999
httplib2: 0.9.2
apiclient: 1.6.5
sqlalchemy: 1.1.3
pymysql: 0.6.1.None
psycopg2: None
jinja2: 2.8
boto: 2.48.0
pandas_datareader: 0.2.0

</details>

### PR DESCRIPTION
**Pull request by attack68**
- [x] closes #21266 
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

This is a draft PR to address 21266: Tooltips for DataFrames in Styler.

This PR has the following objectives:

1) Add functionality whilst being completely (almost) decoupled from the rest of the HTML rendering process
2) Be fully backwards compatible and have no impact at all on previous Styler's that have never used tooltips.
3) Use pseudo CSS class rendering via `table_styles` to control the visualisation of the tooltips.

To address 1) the architecture was simple:

- If `tooltips` are detected then an additional HTML element is added to every data cell: `<span class="pd-t"></span>`. If `tooltips` are not detected (by default) nothing is done.
- Add generic table level CSS for the tooltip class, hiding tooltips by default but positioning, sizing, and coloring them.
- Loop through a tooltips DataFrame and based on the row and col index add additional content to the <span> element using the ::after pseudo-element targeting the exact cell id. This is table level also. Rendering is performed as the last step.

To address 2) default values ensure the functions make no changes unless `set_tooltips` has been called.

## Extensions

The very simple architecture requires a DataFrame of tooltips to be supplied. This requires the user to have constructed it privately (possibly using the regular DataFrame.apply and DataFrame.applymap) methods. These could be incorporated.

It would also be better to only add extra HTML <span> elements to data cells that have tooltips. This requires conditional looping of the render dict, which I haven't included here to save on initial complexity. For small dataframes it is also irrelevant.

However, given my time constraints I would rather leave this to a more available and more competent developer!!

Comments welcome..

### CODE REVIEW

File: pandas/tests/io/formats/test_style.py [hunk 1]
```diff
@@ -1688,6 +1688,34 @@ def test_no_cell_ids(self):
...
+            "#T__ #T__row0_col0:hover .pd-t {\n          visibility: visible;\n    }  "
+            + '  #T__ #T__row0_col0 .pd-t::after {\n          content: "Min";\n    }'
+            in s.render()
+        )
+        # test 'max' tooltip added
+        assert (
+            "#T__ #T__row1_col1:hover .pd-t {\n          visibility: visible;\n    }  "
+            + '  #T__ #T__row1_col1 .pd-t::after {\n          content: "Max";\n    }'
+            in s.render()
+        )
```
Comment (attack68) [inline pandas/tests/io/formats/test_style.py:1711]: Should probably add the fact that ´<span class="pd-t"></span>´ is added to the relevant <td> objects.

File: pandas/tests/io/formats/test_style.py [hunk 2]
```diff
@@ -1688,6 +1688,34 @@ def test_no_cell_ids(self):
...
+
+    def test_tooltip_ignored(self):
+        # GH XXXXX
+        df = pd.DataFrame(data=[[0, 1], [2, 3]])
+        s = Styler(df, uuid="_")  # no set_tooltips()
+        assert '<style  type="text/css" >\n</style>' in s.render()
```
Comment (attack68) [inline pandas/tests/io/formats/test_style.py:1717]: Should probably add that no ´<span class="pd-t"></span>´ are generated in this case.

**Comment by attack68** [comment]
@reviewers: this is ready to be reviewed. All green, except for known unrelated issues

**Comment by attack68** [comment]
@TomAugspurger @WillAyd  you both commented on the underlying issue. Would you care to review this pull request for suitability?

**Comment by TomAugspurger** [comment]
It'll be a while before I can look. No need to hold it up for me.

File: doc/source/user_guide/style.ipynb [hunk 3]
```diff
@@ -1241,4 +1241,4 @@
  },
  "nbformat": 4,
  "nbformat_minor": 1
-}
+}
```
Comment (jreback) [inline doc/source/user_guide/style.ipynb:1244]: can you revert this

File: pandas/io/formats/style.py [hunk 4]
```diff
@@ -159,7 +159,7 @@ def __init__(
         self.index = data.index
         self.columns = data.columns
 
-        self.uuid = uuid
+        self.uuid = uuid or (uuid4().hex[:5] + "_")
```
Comment (jreback) [inline pandas/io/formats/style.py:162]: not averse, but what's the reason for this change?

File: pandas/io/formats/style.py [hunk 5]
```diff
@@ -171,6 +171,11 @@ def __init__(
         self.cell_ids = cell_ids
         self.na_rep = na_rep
 
+        self.tooltip_styles: List[Dict[str, object]] = []  # VERSION ADDED 1.X
```
Comment (jreback) [inline pandas/io/formats/style.py:174]: you don't need the comment

File: pandas/io/formats/style.py [hunk 6]
```diff
@@ -802,6 +810,146 @@ def where(
             lambda val: value if cond(val) else other, subset=subset, **kwargs
         )
 
+    def set_tooltips(self, ttips: DataFrame):
```
Comment (jreback) [inline pandas/io/formats/style.py:813]: can you add the return type (Styler)

File: pandas/io/formats/style.py [hunk 7]
```diff
@@ -802,6 +810,146 @@ def where(
...
+        """
+        Add string based tooltips that will appear in the `Styler` HTML result.
+
+        Parameters
+        ----------
+        ttips : DataFrame
+            DataFrame containing strings that will be translated to tooltips. Empty
```
Comment (jreback) [inline pandas/io/formats/style.py:820]: why ignore empty strings? (we don't normally treat these as nulls)

File: pandas/io/formats/style.py [hunk 8]
```diff
@@ -802,6 +810,146 @@ def where(
...
+        self : Styler
+
+        Notes
+        -----
+        Tooltips are created by adding `<span class="pd-t"></span>` to each data cell
+        and then manipulating the table level CSS to attach pseudo hover and pseudo
+        after selectors to produce the required the results. For styling control
+        see `:meth:Styler.set_tooltips_class`.
+        Tooltips are not designed to be efficient, and can add large amounts of
+        additional HTML for larger tables, since they also require that `cell_ids`
```
Comment (jreback) [inline pandas/io/formats/style.py:835]: can you add an Example here

File: pandas/io/formats/style.py [hunk 9]
```diff
@@ -802,6 +810,146 @@ def where(
...
+        is forced to `True`.
+        """
+        if not (self.columns.equals(ttips.columns) and self.index.equals(ttips.index)):
+            raise AttributeError(
```
Comment (jreback) [inline pandas/io/formats/style.py:839]: make this a ValueError

File: pandas/io/formats/style.py [hunk 10]
```diff
@@ -802,6 +810,146 @@ def where(
...
+                "Tooltips DataFrame must have identical column and index labelling "
+                "to underlying."
+            )
+
+        self.cell_ids = True  # tooltips only work with individual cell_ids
```
Comment (jreback) [inline pandas/io/formats/style.py:844]: isn't this an error if self.cell_ids is not None  (already)?

File: pandas/io/formats/style.py [hunk 11]
```diff
@@ -802,6 +810,146 @@ def where(
...
+        self.tooltip_styles = []
+        for i, rn in enumerate(ttips.index):
+            for j, cn in enumerate(ttips.columns):
+                if ttips.iloc[i, j] in [np.nan, "", None]:
```
Comment (jreback) [inline pandas/io/formats/style.py:848]: do 

```
# in the top
mask = ttips.isna()
```

File: pandas/io/formats/style.py [hunk 12]
```diff
@@ -802,6 +810,146 @@ def where(
...
+                    continue
+                else:
+                    # add pseudo-class and pseudo-elements to css to create tips
+                    self.tooltip_styles.extend(
```
Comment (jreback) [inline pandas/io/formats/style.py:852]: you can actually do this in a list-comprehension

File: pandas/io/formats/style.py [hunk 13]
```diff
@@ -802,6 +810,146 @@ def where(
...
+                                + str(j)
+                                + f" .{self.tooltip_class}::after",
+                                "props": [("content", f'"{str(ttips.iloc[i, j])}"')],
+                            },
+                        ]
+                    )
+
+        return self
+
+    def set_tooltip_class(self, name="pd-t", properties=None):
```
Comment (jreback) [inline pandas/io/formats/style.py:879]: can you add types

**Review by jreback — CHANGES_REQUESTED** [review]
looks reasonable, some comments

File: pandas/io/formats/style.py (hunk 4 above)
Comment (attack68) [reply pandas/io/formats/style.py:162]: a) the uuid is created but later in the process, so **I changed** to bring this forward to instantiation for use with tooltips.
b) uuid1 derives using network address, so **I changed** to use uuid4 which doesn't so is potentially more network secure.
c) a resonable use case is transmitting styled dataframes to web. For large tables without setting `cell_ids` to false the uuid is transmitted at least once and potentially twice for every cell. This restricts it to 5 characters as opposed to many more so compresses the data transfer. The underscore improves HTML readability. I think the loss of randomness is a fair tradeoff?

File: pandas/io/formats/style.py (hunk 7 above)
Comment (attack68) [reply pandas/io/formats/style.py:820]: Allowing empty string as valid will produce a tooltip on hover that doesn't give any additional information for the cell. It strikes me this will be more annoying than making sure it is removed - cant really imagine anyone wanting to create such a tooltip? However, happy to change if needed...

File: pandas/io/formats/style.py (hunk 10 above)
Comment (attack68) [reply pandas/io/formats/style.py]: `cell_ids` is True by default. If the user specifically changes `cell_ids` to False, maybe it is better to raise NotImplementedError for tooltips since the current code doesn't perform optimisation by cell (although is possible).  (see the commit..)

File: doc/source/user_guide/style.ipynb (hunk 3 above)
Comment (attack68) [reply doc/source/user_guide/style.ipynb:1244]: i've no real experience with selected git revert, doesn't it just get squashed?

**Comment by attack68** [comment]
I changed to use reindex for the dataframe shapes similar to #36159 and the comment from @TomAugspurger there. Added test cases.

File: pandas/io/formats/style.py (hunk 4 above)
Comment (attack68) [reply pandas/io/formats/style.py]: I reverted this. Will add it in another PR including tests.

File: pandas/io/formats/style.py [hunk 14]
```diff
@@ -171,6 +171,13 @@ def __init__(
         self.cell_ids = cell_ids
         self.na_rep = na_rep
 
+        self.tooltip_styles: List[Dict[str, object]] = []
```
Comment (jreback) [inline pandas/io/formats/style.py:174]: does it make sense to make this a named tuple? (or a class/dataclass) for clarity here?

File: pandas/io/formats/style.py [hunk 15]
```diff
@@ -545,6 +552,7 @@ def render(self, **kwargs) -> str:
         # so we have the nested anys below
         trimmed = [x for x in d["cellstyle"] if any(any(y) for y in x["props"])]
         d["cellstyle"] = trimmed
+        self._render_tooltips(d)
```
Comment (jreback) [inline pandas/io/formats/style.py:555]: why wouldn't this happen *after* the update?

File: pandas/io/formats/style.py [hunk 16]
```diff
@@ -802,6 +812,158 @@ def where(
...
+        """
+        if not self.cell_ids:
+            # tooltips not optimised for individual cell check.
+            raise NotImplementedError(
+                "Tooltips can only render with 'cell_ids' is True."
+            )
+
+        ttips = ttips.reindex_like(self.data)
+
+        mask = (ttips.isna()) | (ttips.eq(""))
```
Comment (jreback) [inline pandas/io/formats/style.py:858]: is there a reason you are masking with empty string? this is not typical

File: pandas/io/formats/style.py [hunk 17]
```diff
@@ -802,6 +812,158 @@ def where(
...
+        self.tooltip_styles = [
+            d
+            for sublist in [
+                [
+                    {
+                        "selector": "#T_"
```
Comment (jreback) [inline pandas/io/formats/style.py:864]: does it make sense to have the formatter in an function defined in this method? its *very* hard to read / understand the parameters otherwise

File: pandas/io/formats/style.py [hunk 18]
```diff
@@ -802,6 +812,158 @@ def where(
...
+        ...     ('z-index', 1)])
+        """
+        if properties is None:
+            properties = [  # set default
+                ("visibility", "hidden"),
+                ("position", "absolute"),
+                ("z-index", 1),
+                ("background-color", "black"),
+                ("color", "white"),
+                ("transform", "translate(-20px, -20px)"),
```
Comment (jreback) [inline pandas/io/formats/style.py:935]: do we validate the properties? should we?

File: pandas/io/formats/style.py [hunk 19]
```diff
@@ -802,6 +812,158 @@ def where(
...
+            ]
+        self.tooltip_class = name
+
+        self.tooltip_class_styles = [
+            {"selector": f".{self.tooltip_class}", "props": properties}
+        ]
+        return self
+
+    def _render_tooltips(self, d):
```
Comment (jreback) [inline pandas/io/formats/style.py:944]: does this make sense as a module level free function (where you pass in the styles & d)?

File: pandas/tests/io/formats/test_style.py [hunk 20]
```diff
@@ -1697,6 +1697,56 @@ def test_colspan_w3(self):
...
+            ),
+        ],
+    )
+    def test_tooltip_render(self, ttips):
+        # GH 21266
+        df = pd.DataFrame(data=[[0, 3], [1, 2]], columns=["A", "B"], index=["a", "b"])
+        s = Styler(df, uuid="_").set_tooltips(ttips).render()
+        # test tooltip table level class
+        assert "#T__ .pd-t {\n          visibility: hidden;\n" in s
+        # test 'Min' tooltip added
```
Comment (jreback) [inline pandas/tests/io/formats/test_style.py:1720]: blank lines between cases

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/io/formats/style.py [hunk 21]
```diff
@@ -540,6 +627,7 @@ def render(self, **kwargs) -> str:
         self._compute()
         # TODO: namespace all the pandas keys
         d = self._translate()
+        d = self.tooltips._translate_tooltips(self.data, self.uuid, d)
         # filter out empty styles, every cell will have a class
```
Comment (attack68) [inline pandas/io/formats/style.py:631]: this is best before the `update()`, because it is a direct extension of the previous _translate() method.

File: pandas/io/formats/style.py [hunk 22]
```diff
@@ -1516,6 +1605,148 @@ def pipe(self, func: Callable, *args, **kwargs):
...
+                    + "row"
+                    + str(row)
+                    + "_col"
+                    + str(col)
+                    + f" .{name}::after",
+                    "props": [("content", f'"{text}"')],
+                },
+            ]
+
+        mask = (self.tt_data.isna()) | (self.tt_data.eq(""))  # empty string = no ttip
```
Comment (attack68) [inline pandas/io/formats/style.py:1724]: empty strings are removed because otherwise a tooltip would appear as a black (empty) box in HTML on hover, so would just be annoying and useless.

File: pandas/tests/io/formats/test_style.py (hunk 20 above)
Comment (attack68) [reply pandas/tests/io/formats/test_style.py:1722]: done

File: pandas/io/formats/style.py [hunk 23]
```diff
@@ -540,6 +627,7 @@ def render(self, **kwargs) -> str:
         self._compute()
         # TODO: namespace all the pandas keys
         d = self._translate()
+        d = self.tooltips._translate_tooltips(self.data, self.uuid, d)
```
Comment (attack68) [inline pandas/io/formats/style.py:630]: this is best before the update(), because it is a direct extension of the previous _translate() method.

File: pandas/io/formats/style.py [hunk 24]
```diff
@@ -1516,6 +1605,148 @@ def pipe(self, func: Callable, *args, **kwargs):
...
+
+    def __init__(
+        self,
+        css_props: List[Tuple[str, Union[str, int, float]]] = [
+            ("visibility", "hidden"),
+            ("position", "absolute"),
+            ("z-index", 1),
+            ("background-color", "black"),
+            ("color", "white"),
+            ("transform", "translate(-20px, -20px)"),
```
Comment (attack68) [inline pandas/io/formats/style.py:1646]: I do not think there is any need to validate these or manual user properties. CSS is a forgiving language and simply ignores CSS styles if they are not expressed properly. This becomes part of the CSS programmers typical debug pattern.

File: pandas/io/formats/style.py (hunk 14 above)
Comment (attack68) [reply pandas/io/formats/style.py]: I have re-written this PR with a separate class approach, to aid with other comments also...

File: pandas/io/formats/style.py (hunk 15 above)
Comment (attack68) [reply pandas/io/formats/style.py]: becuase it is really an extension of the _translate method.

File: pandas/io/formats/style.py (hunk 16 above)
Comment (attack68) [reply pandas/io/formats/style.py]: yes. For tooltip on hover we could either get a black box pop without text, or no pop up at all. I think the latter is preferred: pop ups should only have useful info. So I filter out.

File: pandas/io/formats/style.py (hunk 17 above)
Comment (attack68) [reply pandas/io/formats/style.py]: tried to write a inner method to explain this.

File: doc/source/user_guide/style.ipynb (hunk 3 above)
Comment (WillAyd) [reply doc/source/user_guide/style.ipynb:1244]: `git checkout upstream/master doc/source/user_guide/style.ipynb` and commit the result

File: pandas/io/formats/style.py [hunk 25]
```diff
@@ -204,6 +206,91 @@ def _repr_html_(self) -> str:
...
+        additional HTML for larger tables, since they also require that `cell_ids`
+        is forced to `True`.
+
+        Examples
+        --------
+        >>> df = pd.DataFrame(data=[[0, 1], [2, 3]])
+        >>> ttips = pd.DataFrame(
+        ...    data=[["Min", ""], [np.nan, "Max"]], columns=df.columns, index=df.index
+        ... )
+        >>> s = Styler(df, uuid="_").set_tooltips(ttips).render()
```
Comment (WillAyd) [inline pandas/io/formats/style.py:242]: Is there a reason to require the uuid argument here? Not immediately obvious as a reader

File: pandas/io/formats/style.py [hunk 26]
```diff
@@ -204,6 +206,91 @@ def _repr_html_(self) -> str:
...
+            raise NotImplementedError(
+                "Tooltips can only render with 'cell_ids' is True."
+            )
+        self.tooltips.tt_data = ttips
+        return self
+
+    def set_tooltips_class(
+        self,
+        name: Optional[str] = None,
+        properties: Optional[List[Tuple[str, Union[str, int, float]]]] = None,
```
Comment (WillAyd) [inline pandas/io/formats/style.py:255]:
```suggestion
        properties: Optional[Sequence[Tuple[str, Union[str, int, float]]]] = None,
```

Or does this have to be a List?

File: pandas/io/formats/style.py [hunk 27]
```diff
@@ -1589,6 +1678,150 @@ def pipe(self, func: Callable, *args, **kwargs):
...
+        - color: white
+        - transform: translate(-20px, -20px)
+
+    Hidden visibility is a key prerequisite to the hover functionality, and should
+    always be included in any manual properties specification.
+    """
+
+    def __init__(
+        self,
+        css_props: List[Tuple[str, Union[str, int, float]]] = [
```
Comment (WillAyd) [inline pandas/io/formats/style.py:1713]: Does this need to be a list? Would be preferable to have a non-mutable default argument

File: pandas/io/formats/style.py [hunk 28]
```diff
@@ -1589,6 +1678,150 @@ def pipe(self, func: Callable, *args, **kwargs):
...
+        """
+        self.tt_data = (
+            self.tt_data.reindex_like(styler_data)
+            .dropna(how="all", axis=0)
+            .dropna(how="all", axis=1)
+        )
+        if self.tt_data.empty:
+            return d
+
+        def pseudo_css(uuid, row, col, name, text):
```
Comment (WillAyd) [inline pandas/io/formats/style.py:1758]: Can you add types for these parameters?

File: pandas/io/formats/style.py [hunk 29]
```diff
@@ -1589,6 +1678,150 @@ def pipe(self, func: Callable, *args, **kwargs):
...
+        self.class_name = css_name
+        self.class_properties = css_props
+        self.tt_data = tooltips
+        self.table_styles: List[Dict[str, Union[str, List[Tuple[str, str]]]]] = []
+
+    @property
+    def class_styles(self):
+        return [{"selector": f".{self.class_name}", "props": self.class_properties}]
+
+    def _translate_tooltips(self, styler_data, uuid, d):
```
Comment (WillAyd) [inline pandas/io/formats/style.py:1733]: Can you annotate this?

**Review by WillAyd — CHANGES_REQUESTED** [review]

File: pandas/io/formats/style.py (hunk 26 above)
Comment (attack68) [reply pandas/io/formats/style.py:255]: The underlying `Styler` `table_styles` get extended with these `_Tooltips` `table_styles` so I have maintained the same type definition for consistency.

Although the `extend()` method would work on sequence as well. Happy to take your lead on this one... Shall I change to sequence?

File: pandas/io/formats/style.py (hunk 27 above)
Comment (attack68) [reply pandas/io/formats/style.py:1713]: Same reason as above: List is for consistency with with underlying `Styler` `table_styles`.

Again shall I change? I probably prefer as is.. fits with existing documentation and other Styler methods.

File: pandas/io/formats/style.py (hunk 25 above)
Comment (attack68) [reply pandas/io/formats/style.py:242]: no. removed. on auto pilot from the other tests I wrote with same signature where it was necessary.

File: pandas/io/formats/style.py (hunk 26 above)
Comment (WillAyd) [reply pandas/io/formats/style.py:255]: Yea let's change to Sequence - typically our API is typed as permissive as possible when it comes to container objects

File: pandas/io/formats/style.py (hunk 27 above)
Comment (WillAyd) [reply pandas/io/formats/style.py:1713]: Hmm fair enough. I'm being pedantic but given your callout let's leave as is and in the future might want to comprehensively change these to non-mutable default arguments

File: pandas/io/formats/style.py [hunk 30]
```diff
@@ -1589,6 +1688,170 @@ def pipe(self, func: Callable, *args, **kwargs):
...
+                    "props": [("content", f'"{text}"')],
+                },
+            ]
+
+        mask = (self.tt_data.isna()) | (self.tt_data.eq(""))  # empty string = no ttip
+        self.table_styles = [
+            style
+            for sublist in [
+                _pseudo_css(i, j, str(self.tt_data.iloc[i, j]))
+                for i, rn in enumerate(self.tt_data.index)
```
Comment (WillAyd) [inline pandas/io/formats/style.py:1834]: Are rn / cn on the next line even used? If not should skip enumerate and just iterate over the length

File: pandas/io/formats/style.py [hunk 31]
```diff
@@ -1589,6 +1688,170 @@ def pipe(self, func: Callable, *args, **kwargs):
...
+                    + str(row)
+                    + "_col"
+                    + str(col)
+                    + f" .{name}::after",
+                    "props": [("content", f'"{text}"')],
+                },
+            ]
+
+        mask = (self.tt_data.isna()) | (self.tt_data.eq(""))  # empty string = no ttip
+        self.table_styles = [
```
Comment (WillAyd) [inline pandas/io/formats/style.py:1830]: Can you just break this into separate for loops / functions? It's a little hard to grok what is going on here with the deeply nested comprehensions

**Review by WillAyd — CHANGES_REQUESTED** [review]

File: pandas/io/formats/style.py (hunk 31 above)
Comment (attack68) [reply pandas/io/formats/style.py:1830]: The objective is for `Styler.table_styles` to be extended by `_Tooltip.table_styles`.
But each data cell (if has tooltip) needs two table styles, so the method `_pseudo_css` generates the list of 2 items for each data cell.
The nesting is only needed to flatten the list to allow the above objective in the right format: `[[a1,b1],[a2,b2]] -> [a1, b1, a2, b2]`.

Happy to take suggestions on structure, but previous versions have had requests for list comprehensions as opposed to separate loops.

**Comment by attack68** [comment]
@jreback @WillAyd this is still live, but it needs agreement on whether you want list comprehensions (previous change request) or not (new request), style wise.

to be honest I think the latest version (jreback's request) is cleanest.

**Comment by hamishdarbyshire** [comment]
would be good to to have the 1.2 milestone on this and then I can work on improving the documentation for all the Styler enhancements I have contributed in subsequent minor versions..

@jreback

**Comment by attack68** [comment]
apologies for posting with wrong account above: work not personal.

**Comment by attack68** [comment]
@WillAyd @jreback can this get a 1.2 milestone tag - all my other Styler PRs made it in and it would be great if this last one could be added.

I can then update the documentation notebook with examples of all new features.

File: pandas/io/formats/style.py [hunk 32]
```diff
@@ -204,6 +206,101 @@ def _repr_html_(self) -> str:
         """
         return self.render()
 
+    def set_tooltips(self, ttips: DataFrame) -> "Styler":
```
Comment (jreback) [inline pandas/io/formats/style.py:209]: can you add an entry for this in style.rst

File: pandas/io/formats/style.py [hunk 33]
```diff
@@ -204,6 +206,101 @@ def _repr_html_(self) -> str:
...
+        """
+        Add string based tooltips that will appear in the `Styler` HTML result. These
```
Comment (jreback) [inline pandas/io/formats/style.py:211]: versionadded 1.2

File: pandas/io/formats/style.py [hunk 34]
```diff
@@ -181,6 +181,8 @@ def __init__(
         self.cell_ids = cell_ids
         self.na_rep = na_rep
 
+        self.tooltips = _Tooltips()
```
Comment (jreback) [inline pandas/io/formats/style.py:184]: shouldn't this just be None by default? and if set_toolstips is called *then* it has a value)

File: pandas/io/formats/style.py (hunk 23 above)
Comment (jreback) [reply pandas/io/formats/style.py:716]: this is *very* confusing  here, why is this not part of _translate itself?

File: pandas/io/formats/style.py [hunk 35]
```diff
@@ -685,6 +783,7 @@ def clear(self) -> None:
         Returns None.
         """
         self.ctx.clear()
+        self.tooltips = _Tooltips()
```
Comment (jreback) [inline pandas/io/formats/style.py:786]: same

File: pandas/io/formats/style.py [hunk 36]
```diff
@@ -1589,6 +1688,170 @@ def pipe(self, func: Callable, *args, **kwargs):
...
+        required to extend the underlying ``Styler`` `table_styles` to allow
+        tooltips to render in HTML.
+
+        Returns
+        -------
+        styles : List
+        """
+        return [{"selector": f".{self.class_name}", "props": self.class_properties}]
+
+    def _translate_tooltips(self, styler_data, uuid, d):
```
Comment (jreback) [inline pandas/io/formats/style.py:1752]: why is this called _translate_tooltips and not _translate?

type the args & return

File: pandas/io/formats/style.py [hunk 37]
```diff
@@ -1589,6 +1688,170 @@ def pipe(self, func: Callable, *args, **kwargs):
...
+            self.tt_data.reindex_like(styler_data)
+            .dropna(how="all", axis=0)
+            .dropna(how="all", axis=1)
+        )
+        if self.tt_data.empty:
+            return d
+
+        name = self.class_name
+
+        def _pseudo_css(row: int, col: int, text: str):
```
Comment (jreback) [inline pandas/io/formats/style.py:1779]: make this a module level function, its *very* confusing here (yes i know its only used here) but this nesting maybe nice in javascript but it precludes readbility

File: pandas/io/formats/style.py [hunk 38]
```diff
@@ -1589,6 +1688,170 @@ def pipe(self, func: Callable, *args, **kwargs):
...
+            for sublist in [
+                _pseudo_css(i, j, str(self.tt_data.iloc[i, j]))
+                for i in range(len(self.tt_data.index))
+                for j in range(len(self.tt_data.columns))
+                if not mask.iloc[i, j]
+            ]
+            for style in sublist
+        ]
+
+        if self.table_styles:
```
Comment (jreback) [inline pandas/io/formats/style.py:1841]: i am puzzled by this here why this is involved at all

**Review by jreback — CHANGES_REQUESTED** [review]
i know you really want this for 1.2, but it needs some work, if you can turn it around shortly then we can do

File: pandas/io/formats/style.py (hunk 32 above)
Comment (attack68) [reply pandas/io/formats/style.py:222]: done, and `set_tooltips_class` also

File: pandas/io/formats/style.py (hunk 33 above)
Comment (attack68) [reply pandas/io/formats/style.py:224]: done, except I guess this is 1.3.0 now? Sorry, wasn't available over the holidays so couldn't get to this.

File: pandas/io/formats/style.py (hunk 34 above)
Comment (attack68) [reply pandas/io/formats/style.py]: I have changed to your preference. Was that structure originally to avoid some mypy issues which I have worked around and since the initialised `_Tooltips()` instance was pretty empty anyway.

File: pandas/io/formats/style.py (hunk 35 above)
Comment (attack68) [reply pandas/io/formats/style.py]: changed to None to your preference.

File: pandas/io/formats/style.py (hunk 37 above)
Comment (attack68) [reply pandas/io/formats/style.py]: changed.

File: pandas/io/formats/style.py (hunk 38 above)
Comment (attack68) [reply pandas/io/formats/style.py:1932]: the above `table_styles` will be an empty list if there are no tooltips to add, i.e if the `mask` was empty. In that case this `if` check provides some efficiency by avoiding adding a `<span>` html selector to every table cell.

File: pandas/io/formats/style.py (hunk 36 above)
Comment (attack68) [reply pandas/io/formats/style.py]: changed. it was legacy name. it was separated within the `render` function because I preffered the decoupled structure. However, I integrated it into the `Styler._translate` method as per your preference.

File: pandas/io/formats/style.py (hunk 33 above)
Comment (jreback) [reply pandas/io/formats/style.py:224]: yes

Comment (jreback) [reply pandas/io/formats/style.py:224]: no problem at all. thanks for the patch.

File: pandas/io/formats/style.py [hunk 39]
```diff
@@ -688,6 +805,7 @@ def clear(self) -> None:
         Returns None.
         """
         self.ctx.clear()
+        self.tooltips = None
```
Comment (jreback) [inline pandas/io/formats/style.py:808]: should be typed the same as above

File: pandas/io/formats/style.py [hunk 40]
```diff
@@ -1657,6 +1775,178 @@ def pipe(self, func: Callable, *args, **kwargs):
...
+            ("color", "white"),
+            ("transform", "translate(-20px, -20px)"),
+        ],
+        css_name: str = "pd-t",
+        tooltips: DataFrame = DataFrame(),
+    ):
+        self.class_name = css_name
+        self.class_properties = css_props
+        self.tt_data = tooltips
+        self.table_styles: List[Dict[str, Union[str, List[Tuple[str, str]]]]] = []
```
Comment (jreback) [inline pandas/io/formats/style.py:1824]: if this is re-used anywhere make an alias at the top of the file

File: pandas/io/formats/style.py (hunk 38 above)
Comment (jreback) [reply pandas/io/formats/style.py:1936]: ok can you add that as a comment

**Review by jreback — CHANGES_REQUESTED** [review]
looks good, some small comments. ping on green.

assume that you will do a followon to add docs to the style notebook? (can also be in this PR)

File: pandas/io/formats/style.py (hunk 39 above)
Comment (attack68) [reply pandas/io/formats/style.py]: done

File: pandas/io/formats/style.py (hunk 40 above)
Comment (jreback) [reply pandas/io/formats/style.py:1825]: is this used anywhere else?

File: pandas/io/formats/style.py (hunk 38 above)
Comment (jreback) [reply pandas/io/formats/style.py:1937]: can you update this

**Review by jreback — CHANGES_REQUESTED** [review]
looks good a question and comment.

File: pandas/io/formats/style.py (hunk 40 above)
Comment (attack68) [reply pandas/io/formats/style.py:1825]: no, did a search, this typing only used here

File: pandas/io/formats/style.py (hunk 38 above)
Comment (attack68) [reply pandas/io/formats/style.py:1937]: i added a comment line 1938?

**Comment by jreback** [comment]
ok @attack68 can you resolve this precommit issue: 
```pandas/io/formats/style.py:809: error: Attribute 'tooltips' already defined on line 185  [no-redef]```

merge master and ping on green-ish

**Comment by attack68** [comment]
@jreback fixed, green-ish... ping.

**Review by jreback — APPROVED** [review]
thanks @attack68 

would not object to a PR to split up pandas/io/formats/style.py to submodules in the future as this file is getting kind of long / including multiple functionaility.

**Comment by simonjayhawkins** [comment]
@jreback this broke pre-commit on master. will fix up shortly so feel free to carry on merging PRs to master

**Comment by jreback** [comment]
thanks @simonjayhawkins
